hmis external referrals 184843920-184816343

ttoomey opened the pull request:
### Please squash merge this PR

[#184816343](https://www.pivotaltracker.com/story/show/184816343)
* external referral models 

[#184843920](https://www.pivotaltracker.com/story/show/184843920)
* send referral request job
* void referral request job
* update referral posting job
* create referral endpoint

misc: 
* upgrade brakeman so it doesn't bomb on ruby 3 syntax
* extract external ids into rails drivers extensions
* use webmock to stop our tests from hitting the pwned password api

out of scope for this pr:
* validation [#184931194](https://www.pivotaltracker.com/story/show/184931194)
* API authentication mechanism
* Mapping of household member demographics to hud client
* Data quality on household members

notes:
* External endpoints are not available at time of writing
* I organized the api interactions into jobs since they are a convenient service-object shape. It's possible that some of them could run async but that's for the future

eanders commented on app/models/grda_warehouse/external_referral.rb:
```diff
@@ -0,0 +1,13 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module GrdaWarehouse
+  # A request for a service for a household. The service is not necessarily housing.
+  class ExternalReferral < GrdaWarehouseBase
+    has_many :external_referral_clients, class_name: 'GrdaWarehouse::ExternalReferralClient', dependent: :destroy
+    has_many :external_referral_postings, class_name: 'GrdaWarehouse::ExternalReferralPosting', dependent: :destroy
+  end
+end
```
@ttoomey these should all live underneath `/drivers/hmis/app/models` (unless @gigxz had a reason for them to be out here.)

eanders replied:
@gigxz, were you thinking these should be in their own driver maybe?

ttoomey replied:
Thanks! My mistake, I'll move them to the hmis namespace

eanders replied:
Feel free to hold off on the move until we hear from @gigxz next week, I'm not sure if we should compartmentalize this in its own driver or not.

gigxz replied:
Good question. I was thinking all the External* models would go in `drivers/hmis_external_apis/app/models` which @yakloinsteak is creating here https://github.com/greenriver/hmis-warehouse/pull/2933. The UnitType model would go in `drivers/hmis/app/models`. @eanders what do you think of that? I think I'd prefer the HMIS driver to contain only the "core" models, and the external driver can contain as much as possible of the external business logic (some of which may be on the models).

@ttoomey for now, you can just move these to `/drivers/hmis/app/models`, and I can move them over to the external driver once it's set up (unless Elliot objects). thanks!

eanders replied:
That sounds like a plan to me.  Core HMIS (HUD compliant bits and other things that will be enabled in all installations in `drivers/hmis` HMIS specific API infrastructure in `drivers/hmis_external_apis`.

ttoomey replied:
@gigxz Thanks! I'll just move external_* models hmis_external_apis right now and steal the relevant bits from #2933. I'd prefer to move them now before I build on them. Sound okay?

gigxz replied:
Sounds great, thanks

ttoomey replied:
@gigxz @eanders Any objection to using model names like this: `HmisExternalApis::Referral`, `HmisExternalApis::ReferralRequest`, etc ?

And respective table names  would be `hmis_external_referrals`, `hmis_external_referral_requests`. Or is it convention that the table prefix match the driver namespace (so `hmis_external_api_referrals`)?

eanders replied:
Those names are good.  No need to include the driver namespace in the db at this time.

eanders commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,52 @@
+class AddExternalReferrals < ActiveRecord::Migration[6.1]
+  def change
+    create_table :unit_types do |t|
+      t.timestamps
+      t.string :description
+      t.integer :bed_type # HUD bed type
+      t.integer :unit_size
+      # checking... this would be external_id_id. Is this right?
+      t.references :external_id
+    end
```
@gigxz are these specific to referrals or will we use the same unit types for "standard" units?

gigxz replied:
They are not specific to referrals, they would be used for "standard" units too.

At the last meeting I suggested that we use the `external_ids` table for housing external IDs for unit types, too, but you objected to that.. I don't remember why. Do you? Or maybe I misunderstood you.

eanders replied:
Hmm, I don't remember why I objected.  If there's a `unit_types` table and we're getting the unit types from an external source or matching against one, then I think it makes sense.  I think I was thinking these weren't going to saved to a table, just in code, but I like this better, as long as the external reference is optional.

gigxz replied:
Oh wait, looking back at the document, it seems like we _did_ decide that. The unit type table says `has_one :external_id` - that was probably confusing to write, sorry, but the intention was that the external_ids table has records where `source_type: Hmis::UnitType`. That will be a "belongs to" relationship for the unit type. I think you can go ahead and remove this `external_id ` column from the unit types table.

gigxz replied:
I didn't see your message when I was writing my second one, but yeah. It is optional. @ttoomey you can remove this column.

eanders commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,52 @@
+      t.string :description
+      t.integer :bed_type # HUD bed type
+      t.integer :unit_size
+      # checking... this would be external_id_id. Is this right?
```
I think we settled on `external_id.value` for the storage of the external identifier, but yeah, if it's referencing the active record model, then I think it would end up being `external_id_id`.  That begs the question of how tightly tied these should be, maybe it should just be `external_id` which would match `external_ids.value` but require both be obtained through external systems.  (As in you could have an `external_id` here that didn't necessarily exist yet in `external_ids`

ttoomey replied:
removed the external id reference per thread above https://github.com/greenriver/hmis-warehouse/pull/2955#discussion_r1161658441

ttoomey committed:
184843920, 184816343 hmis external referrals

models
send endpoint
fix issue truncating projects during tests
cspell config for vscode extension

ttoomey committed:
move procedure to job

ttoomey committed:
wip

ttoomey committed:
wip

ttoomey committed:
cleanups

gigxz commented on app/models/grda_warehouse/utility.rb:
```diff
@@ -151,6 +151,7 @@ def self.modifier(model)
     cascade_models = [
       GrdaWarehouse::DataSource,
       GrdaWarehouse::Hud::Client,
+      GrdaWarehouse::Hud::Project,
       GrdaWarehouse::ServiceHistoryEnrollment,
     ]
     return 'CASCADE' if cascade_models.include?(model)
```
Would want @eanders to OK this if he hasn't already..

eanders replied:
That should be fine, this is only used in tests and occasionally in dev.  I believe it will require some data that cascades to correctly work though.  (My memory on that is not clear.)

gigxz commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,51 @@
+      t.string :identifier, null: false, index: { unique: true, name: 'uidx_hmis_external_referrals_identifier' }
+      t.date :referral_date, null: false
+      t.string :service_coordinator, null: false
+      t.string :raw_request
```
```suggestion
      t.string :raw_request
```

We can use the external_request_logs table to store raw requests

gigxz commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,51 @@
+    create_table :hmis_external_referral_clients do |t|
+      t.timestamps
+      t.references :referral, null: false, foreign_key: { to_table: :hmis_external_referrals }, index: { name: 'idx_hmis_external_referral_clients_on_referral_id' }
+      t.references :hud_client, null: false, index: false, foreign_key: { to_table: 'Client' }
```
Can we call this column `client_id` ? That's what we do elsewhere in the db when we are referencing the database id. ("hud_client_id" is especially confusing because there _is_ a special HMIS ClientID that is different from the database id –– but that's not what we're referring to here.)

ttoomey replied:
Sure! I'd changed it since we are calling this table "referral_clients" so client.client was a little hard to track. What do you think about renaming this table "hmis_external_referral_household_members"? That's more reflective of what we receive from LINK

gigxz replied:
That makes sense to me

gigxz commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,51 @@
+      t.string :raw_request
+    end
+
+    create_table :hmis_external_referral_clients do |t|
```
This table needs:
```
t.integer :relationship_to_hoh, null: false
```
which will allow us to determine which Client in a given referral is the "head of household."

The mapping of possible values is at `::HudLists.relationship_to_ho_h_map`.

gigxz commented on db/warehouse/migrate/20230407164611_add_external_referrals.rb:
```diff
@@ -0,0 +1,51 @@
+      t.string :identifier, null: false, index: { unique: true, name: 'uidx_hmis_external_referral_posting_identifier' }
+      t.integer :status, null: false
+      t.references :referral, null: false, foreign_key: { to_table: :hmis_external_referrals }, index: false
+      t.references :referral_request, null: false, foreign_key: { to_table: :hmis_external_referral_requests },
+                                      index: { name: 'idx_hmis_external_referral_postings_on_request_id' }
```
This just in: in some cases the LINK will hit the Send Referral endpoint as part an "admin assign" workflow. That means someone(s) is being assigned to a program directly, without linking it to a vacancy request at all.

In that workflow, the Send Referral payload will include the "program ID" (MPER external Project ID) in place of the "referral request ID," because there is no associated referral request. (I suspect they will also need to send Unit Type ID for admin assigns, in which case we will need to add a column here for that as well. I should know by Monday afternoon's meeting.)

So, my suggested changes are:
1. Make the referral_request_id column optional
2. Add a project_id column, and make it required. It means we are storing the project ID redundantly for non-admin-assigned postings, but I think that's ok, since we need that column here anyway.


...A _different_ approach would be to create stub referral requests for admin-assigned referrals that come in. I think that would also be reasonable, but we'd just need to mark them as such and make several of the columns nullable. Let me know if you have a strong feeling either way @ttoomey.

ttoomey replied:
That sounds good, your suggestion makes sense to me. I'll adjust the schema.

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb.rej:
```diff
@@ -0,0 +1,10 @@
+   has_many :files, class_name: '::Hmis::File', dependent: :destroy, inverse_of: :client
+   has_many :current_living_situations, through: :enrollments
+   has_many :hmis_services, through: :enrollments # All services (HUD and Custom)
++  has_many :external_referral_clients, class_name: 'HmisExternalApis::ReferralClient', dependent: :destroy, inverse_of: :hud_client
```
Looks like this change got dropped?

ttoomey replied:
Oh oops, this was some cruft from rebasing into r71. I'll remove it, thanks!

gigxz commented on drivers/hmis/app/models/hmis/hud/project.rb:
```diff
@@ -20,6 +20,7 @@ class Hmis::Hud::Project < Hmis::Hud::Base
   has_many :inventories, **hmis_relation(:ProjectID, 'Inventory'), inverse_of: :project, dependent: :destroy
   has_many :funders, **hmis_relation(:ProjectID, 'Funder'), inverse_of: :project, dependent: :destroy
+  has_many :external_referral_requests, class_name: 'HmisExternalApis::ReferralRequest', dependent: :destroy
 
```
This will need a has_many for Postings, too.

gigxz commented on drivers/hmis/spec/factories/hmis/users.rb:
```diff
@@ -1,10 +1,11 @@
 FactoryBot.define do
   factory :hmis_user, class: 'Hmis::User', parent: :user do
-    data_source { association :hmis_data_source }
+    # FIXME: user.data_source association doesn't exist?
```
Hmm I'm not sure what's going on here, let me know if you want to look together. Majority of tests use `'hmis base setup'` context to set up a data source, user, organization, project, and client.

ttoomey replied:
I see, I should use the hmis_user helper method and not this factory. I get an error when I try and use the factory as-is, for example the following doesn't work:
```ruby
factory :hmis_external_api_referral_request, class: 'HmisExternalApis::ReferralRequest' do
  # error: `Hmis::User.data_source` is not defined 
  association :requested_by, factory: :hmis_user
  # ...
end
```

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+        .pluck(:value, :source_id)
+        # external id values are not unique, pick the first record by id
+        # https://github.com/greenriver/hmis-warehouse/pull/2933/files#r1164302287
+        .sort_by(&:last).reverse.to_h
```
We got an update on this. In the event of multiple Clients with the same MCI ID, should choose the Client with the earliest creation date.

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+###
+
+module HmisExternalApis
+  class CreateReferralJob < ApplicationJob
```
I like the usage of Job so that we can easily make this async later. Good idea.

gigxz commented on drivers/hmis_external_apis/app/controllers/hmis_external_apis/referrals_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    skip_before_action :authenticate_user!
+    prepend_before_action :skip_timeout
+
+    def create
```
We should add in a `ExternalRequestLog` here eventually. What do you think of this as a pattern?
```ruby
log = ExternalRequestLog.create!(<request info>)
begin
   # do the thing
rescue
   # construct error response
end
log.update(response: <response>)
render json: <response>
```

Not necessary to actually do in this PR, I know there are stubs still for the business logic and error handling, so it can go in later with those.

We will need to set up the `warehouse_api_configurations` table so that the `ExternalRequestLog` can reference an `initiator`. (The initiator being our "referrals" api).

ttoomey replied:
So we'd use an exception to handle validation failures? That might be necessary but I suggest we try a normal control flow first. Then exceptions would be reserved for errors and we could just leave them unhandled (ensure instead of rescue). Something like this:

```ruby
    def create
      # with_logging do
      #  ...
      # end
    end

    def with_logging
      log = ExternalRequestLog.create!(<request info>)
      begin
        yield
      ensure
        log.update!(response: <response>)
      end
    end
```

gigxz replied:
Ok, I like that!

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+    protected
+
+    def create_referral
+      params => {referral_id:, referral_date:, service_coordinator:}
```
niiice, I didn't know ruby had destructuring.

ttoomey replied:
I believe destructuring assignment for hash values was introduced in ruby 3. It's new to me too :)

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+        if found_id
+          client.hud_client_id = found_id
+        else
+          client.hud_client = create_hud_client(attrs)
```
Minor thing for readability, but the "hud_client" thing is throwing me off (as mentioned above in the db schema). We don't really use that term elsewhere. I'd prefer:
`referral_client.client = create_client(attrs)`

ttoomey replied:
sure thing, I'll rename it

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+
+      member_params.map do |attrs|
+        attrs => {mci_id:, relationship_to_hoh:}
+        client = referral.referral_clients.new
```
```suggestion
        client = referral.referral_clients.new(relationship_to_hoh: relationship_to_hoh)
```

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+      client = Hmis::Hud::Client.new
+      client.data_source = data_source
+      client.first_name = first_name
+      client.middle_name = last_name
```
```suggestion
      client.middle_name = middle_name
```

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+
+      # DQ = 1 needed for validations
+      client.NameDataQuality = 1
+      client.SSNDataQuality = 1
```
SSN data quality wont necessarily be 1. We can infer it though based on what the ssn looks like. I can update this when I do the demographic mappings.

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb:
```diff
@@ -0,0 +1,140 @@
+      # FIXME: unsure what client.user is and why it's required. Docs needed
+      user = Hmis::Hud::User.new
+      user.data_source = data_source
+      user.save!
```
This is the "HUD user" that most recently updated the record. A "hud user" is a representation of an application user that gets exported as part of the HUD spec.

In this case, we should use the system user. Below is the logic for it - it would be helpful to make a method like `Hmis::Hud::User.system_user(data_source_id)` that returns it:

```
system_user = Hmis::User.find(User.system_user.id)
system_user.hmis_data_source_id = data_source_id
hud_user = Hmis::Hud::User.from_user(system_user)
```

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_request_job.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+
+    def format_date(date)
+      date.strftime('%Y/%m/%d')
```
```suggestion
      date.strftime('%Y-%m-%d')
```

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_request_job.rb:
```diff
@@ -0,0 +1,60 @@
+    # FIXME: not sure best to lookup mper ids... assuming they are external_ids that will already exist?
+    def mper_id(record)
+      mper_cred.external_ids.where(source: record).first!.value
+    end
```
Yep, this is exactly right.

There are other times when we'll want to look up the external ID for a project / organization, so it could also make sense to put something like the Client's `external_ids_by_slug` into a shared concern like `include HasExternalIds`. Then we could just call `project.external_id_by_slug(slug: "mper")`. Not necessary though, what you have here looks correct to me.

gigxz reviewed (commented):
Looks great!! Comments inline.

ttoomey committed:
Merge branch 'release-71' into 184843920-184816343-external-referrals

ttoomey committed:
wip, address PR feedback

ttoomey committed:
brakeman

ttoomey committed:
update referral posting endpoint

ttoomey committed:
void endpoint

ttoomey committed:
Merge branch 'release-71' into 184843920-184816343-external-referrals

ttoomey committed:
suspect we still have ruby 2 in ci

eanders commented on app/models/grda_warehouse/remote_credential.rb:
```diff
@@ -8,9 +8,21 @@ module GrdaWarehouse
   class RemoteCredential < GrdaWarehouseBase
     acts_as_paranoid
     attr_encrypted :password, key: ENV['ENCRYPTION_KEY'][0..31]
+    has_many :external_ids,
+             class_name: 'HmisExternalApis::ExternalId',
+             foreign_key: :remote_credential_id,
+             dependent: :restrict_with_exception
```
This and the class methods below should be injected from the driver via an extension.
See https://github.com/degica/rails_drivers#extensions for details.
`drivers/adult_only_households_sub_pop/extensions/grda_warehouse/hud/client_extension.rb` has a simple example. 
You'll need to include `include RailsDrivers::Extensions` in `remote_credentials.rb`

eanders commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -30,7 +30,8 @@ class Hmis::Hud::Client < Hmis::Hud::Base
   has_many :current_living_situations, through: :enrollments
   has_many :hmis_services, through: :enrollments # All services (HUD and Custom)
-  has_many :external_ids, class_name: 'HmisExternalApis::ExternalId', foreign_key: :source_id
+  has_many :external_ids, class_name: 'HmisExternalApis::ExternalId', as: :source
+  has_many :external_referral_household_members, class_name: 'HmisExternalApis::ReferralHouseholdMember', dependent: :destroy, inverse_of: :client
```
These relationships should be injected via extension from the `hmis_external_apis`

gigxz replied:
Can you say more about this @eanders? If this desire is to have no dependencies from `hmis` on `hmis_external_apis`, I don't think that's feasible. There will be several places where we need to call methods in the external api driver directly: Referral Request form processor, Client form processor (for MCI), initiate unit availability update, etc.

eanders replied:
Sure, this isn't an attempt to completely decouple the drivers, just that the driver providing the has_many should be injecting it.  Calling something can be easily wrapped with an `if driver is installed` later, but this moves the definition of the functionality into the driver where the models live.

gigxz replied:
Oh gotcha. Thanks for the explanation!

eanders commented on drivers/hmis/app/models/hmis/hud/project.rb:
```diff
@@ -20,6 +20,9 @@ class Hmis::Hud::Project < Hmis::Hud::Base
   has_many :project_cocs, **hmis_relation(:ProjectID, 'ProjectCoc'), inverse_of: :project, dependent: :destroy
   has_many :inventories, **hmis_relation(:ProjectID, 'Inventory'), inverse_of: :project, dependent: :destroy
   has_many :funders, **hmis_relation(:ProjectID, 'Funder'), inverse_of: :project, dependent: :destroy
+  has_many :external_ids, class_name: 'HmisExternalApis::ExternalId', as: :source
+  has_many :external_referral_requests, class_name: 'HmisExternalApis::ReferralRequest', dependent: :destroy
+  has_many :external_referral_postings, class_name: 'HmisExternalApis::ReferralPosting', dependent: :destroy
```
These relationships should be injected via extension from the `hmis_external_apis`

eanders commented on drivers/hmis/app/models/hmis/unit_type.rb:
```diff
@@ -0,0 +1,50 @@
+module Hmis
+  class UnitType < HmisBase
+    has_many :external_referral_requests, class_name: 'HmisExternalApis::ReferralRequest', dependent: :restrict_with_exception
+    # has_many :external_referral_postings, class_name: 'HmisExternalApis::ReferralPosting', dependent: :restrict_with_exception
```
These relationships should be injected via extension from the `hmis_external_apis`

ttoomey committed:
probe ci failure

ttoomey committed:
Merge branch 'release-71' into 184843920-184816343-external-referrals

ttoomey committed:
move cross-driver dependencies to extensions

ttoomey committed:
brakeman upgrade, trying to address CI-only failure

eanders commented on Gemfile:
```diff
@@ -237,7 +237,7 @@ group :development, :test do
   gem 'pry-rails'
   gem 'bundler-audit', require: false
-  gem 'brakeman', '~> 5.0.4', require: false
+  gem 'brakeman', require: false
```
Thanks!  I wonder why we'd pinned it.  Probably intended to >= 5.0.4...  This works great though.

eanders commented on config/initializers/inflections.rb:
```diff
@@ -18,6 +18,10 @@
   inflect.irregular 'census', 'censuses'
   inflect.irregular 'supplemental_enrollment_data', 'supplemental_enrollment_data'
 
+  # rails_drivers fails to load extensions for "plural" names since classify() will singularize the driver name
```
Well isn't that special!

eanders reviewed (approved):
Thanks for the refactor!

gigxz commented:
Woohoo thanks @ttoomey!

@eanders  - are you ok with this going in release-71?
(also: Theron requested a squash-merge, which seems like a good idea here)

eanders commented:
Both of those are fine.
Thanks @ttoomey!

gigxz committed:
Merge branch 'release-71' into 184843920-184816343-external-referrals
